### PR TITLE
Implement authorization

### PR DESCRIPTION
**Pull request by GordonOtieno**
- Installed CanCanCan in the project.
- Add a role column to the users table. Used migration for this.
- A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). 
     -  Implemented the post deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it. 
 - A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin"). Use CanCanCan for this authorization.
    -  Implemented the comment-deleting functionality. Add the "Delete" button to the view and make sure that only authorized users can see it.

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -5,9 +5,9 @@ def index
   end
 
   def show
-    @post = Post.includes(:author, :comments, :likes).find(params[:id])
+    @post = Post.find(params[:id])
     @user = User.find(params[:user_id])
-    @comments = @post.comments
+    @comments = @post.comments.includes(:author)
   end
```
Comment (Munsa1) [inline app/controllers/posts_controller.rb:12]: - [ ] It is a requirement that you use `CanCanCan` for Post authorization in `Post Controller`, kindly do so
image

File: app/controllers/comments_controller.rb [hunk 2]
```diff
@@ -9,15 +9,27 @@ def create
...
-      flash[:success] = 'Comment created successfully'
-      redirect_to "/users/#{@current_post.author.id}/posts/#{@current_post.id}"
+      redirect_to user_posts_path(@current_post.author_id, @current_post.id), notice: 'New comment saved successfully'
 
     else
-      flash[:error] = 'Something went wrong'
-      render :new, status: :unprocessable_entity
+      render :new, status: :unprocessable_entity, alert: 'Something went wrong'
     end
   end
```
Comment (Munsa1) [inline app/controllers/comments_controller.rb:18]: - [ ] A user has to be able to add a comment and be able to delete the comment to other users post, kindly implement this feature as it is a reqiurement
image

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @GordonOtieno  @t-bello7 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
:heavy_check_mark: Descriptive Readme
:heavy_check_mark: Passing Linters
:heavy_check_mark: Correct Gitflow
## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (peterrobert) [reply app/controllers/posts_controller.rb:12]: This change has been done well

File: app/views/users/show.html.erb [hunk 3]
```diff
@@ -9,7 +9,7 @@
     <div class="d-flex justify-content-end" style="width: 350px">
       <p class="card-text">
         <small class="text-muted"
-          >Number of posts: <%= @user.postsCounter %></small
+          >Number of posts: <%= @user.posts_counter %></small
```
Comment (peterrobert) [inline app/views/users/show.html.erb]: It seems this change has introduced a bug to your application. When I run the application, it throws an error when I try navigating to the user show page. The reason is that you have mistyped it, it should not be `@user.posts_counter` but it should be `@user.postsCounter`.

- [ ] It will be great to revert the change made here as it is causing a bug to the application

File: app/controllers/comments_controller.rb (hunk 2 above)
Comment (peterrobert) [reply app/controllers/comments_controller.rb:18]: This change has been done well

**Review by peterrobert — CHANGES_REQUESTED** [review]
Hi @GordonOtieno, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏


Feel free to leave any questions or comments in the PR thread or tag me @peterrobert if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have only 3 code reviews per project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by HAMMAS-SALEEM — APPROVED** [review]
# Status: Approved :green_circle: :1st_place_medal: :100:
success

### Hey @GordonOtieno, @t-bello7 👋

Great work on implementing the requested changes. Your project is complete. :blush:
There is nothing else to say about your project. :+1:

## Highlights :1st_place_medal:

✔️ Descriptive PR Title
✔️ Descriptive PR Summary
✔️ Follow Gitflow
✔️ Following Requirements

# Optional

_All the changes with optional tags are not crucial enough to prevent you from getting the approval but we highly recommend to students that they should take these changes into consideration._

Cheers and Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

---------------------------------------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
